File: NinePSharp.Messages.FSharp/messages/Txattrwalk.fs
namespace NinePSharp.Messages.FSharp

open System
open System.Buffers.Binary
open NinePSharp.Constants
open NinePSharp.Interfaces
open NinePSharp.Messages
open NinePSharp.Messages.FSharp.Protocol

type Txattrwalk =
    { Size: uint32
      Tag: uint16
      Fid: uint32
      NewFid: uint32
      Name: string }

    interface ISerializable with
        member this.Size = this.Size
        member this.Type = MessageTypes.Txattrwalk
        member this.Tag = this.Tag
        member this.WriteTo(span: Span<byte>) =
            let mutable offset = 0
            writeHeaders this.Size this.Tag MessageTypes.Txattrwalk span &offset
            BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), this.Fid); offset <- offset + 4
            BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), this.NewFid); offset <- offset + 4
            writeString this.Name span &offset

    static member Parse (data: ReadOnlySpan<byte>) =
        let mutable offset = 0
        let size, tag = readHeaders data &offset
        let fid = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4))
        offset <- offset + 4
        let newFid = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4))
        offset <- offset + 4
        let name = readString data &offset
        { Size = size; Tag = tag; Fid = fid; NewFid = newFid; Name = name }

File: NinePSharp/messages/Txattrwalk.cs
namespace NinePSharp.Messages;

using System;
using System.Buffers.Binary;
using NinePSharp.Constants;
using NinePSharp.Interfaces;
using NinePSharp.Protocol;

public readonly struct Txattrwalk : IMessage
{
    public uint Size { get; }
    public MessageTypes Type => MessageTypes.Txattrwalk;
    public ushort Tag { get; }
    public uint Fid { get; }
    public uint NewFid { get; }
    public string Name { get; }

    public Txattrwalk(uint size, ushort tag, uint fid, uint newFid, string name)
    {
        Size = size;
        Tag = tag;
        Fid = fid;
        NewFid = newFid;
        Name = name;
    }

    public Txattrwalk(ReadOnlySpan<byte> data)
    {
        Size = BinaryPrimitives.ReadUInt32LittleEndian(data[..4]);
        Tag = BinaryPrimitives.ReadUInt16LittleEndian(data.Slice(5, 2));
        int offset = NinePConstants.HeaderSize;
        Fid = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4)); offset += 4;
        NewFid = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4)); offset += 4;
        Name = data.ReadString(ref offset);
    }

    public void WriteTo(Span<byte> span)
    {
        span.WriteHeaders(Size, Tag, MessageTypes.Txattrwalk);
        int offset = NinePConstants.HeaderSize;
        BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), Fid); offset += 4;
        BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), NewFid); offset += 4;
        span.WriteString(Name, ref offset);
    }
}
